### ISSUE TITLE
Vasp Incar check_params() can't handle Union types

### ISSUE THREAD

**Issue by wladerer** [materialsproject/pymatgen#3957]
### Python version

3.11.9

### Pymatgen version

2024.6.10

### Operating system version

Ubuntu 22.04.4 LTS

### Current behavior

Error Message:
```
Traceback (most recent call last):
  File "/home/wladerer/github/abvio/tests/test_incar.py", line 356, in test_valid_model
    incar.check_params()
  File "/home/wladerer/.venv/base/lib/python3.11/site-packages/pymatgen/io/vasp/inputs.py", line 979, in check_params
    if param_type is not None and not isinstance(val, eval(param_type)):
                                                      ^^^^^^^^^^^^^^^^
  File "<string>", line 1, in <module>
NameError: name 'Union' is not defined

----------------------------------------------------------------------
Ran 1 test in 0.002s

FAILED (errors=1)
Finished running tests!
```

Incar.check_params() does not seem to handle Union types

Something strange is happening here where `isinstance(eval(param_type))`
```python
# Check value and its type
            param_type = incar_params[tag].get("type")
            allowed_values = incar_params[tag].get("values")

            if param_type is not None and not isinstance(val, eval(param_type)):
                warnings.warn(f"{tag}: {val} is not a {param_type}", BadIncarWarning, stacklevel=2)
```
Is unaware of the typing module Union type. I can see that it is not imported, so when I update the imports
```python
from typing import TYPE_CHECKING, Any, Literal, Union
```
the error goes away. 

```
test_valid_model (test_incar.TestIncarModel.test_valid_model) ... EDIFF tag type: <class 'str'> tag value <class 'float'>
ISMEAR tag type: <class 'str'> tag value <class 'int'>
LREAL tag type: <class 'str'> tag value <class 'bool'>
EDIFFG tag type: <class 'str'> tag value <class 'float'>
LWAVE tag type: <class 'str'> tag value <class 'bool'>
MAGMOM tag type: <class 'str'> tag value <class 'list'>
ok

----------------------------------------------------------------------
Ran 1 test in 0.001s

OK
Finished running tests!
```

### Expected Behavior

No output 

### Minimal example

```Python
import unittest

from pymatgen.io.vasp import Incar 


class TestCheckIncarParams(unittest.TestCase):

    def test_dict_with_magmoms(self):

        incar_dict = {
            "SYSTEM": "Test",
            "ENCUT": 400,
            "ISMEAR": 0,
            "SIGMA": 0.1,
            "IBRION": 2,
            "NSW": 0,
            "LCHARG": False,
            "LWAVE": False,
            "LORBIT": 11,
            "LREAL": "Auto",
            "ISYM": 0,
            "MAGMOM": [0.6, 0.6, 0.2],
        }

        incar = Incar.from_dict(incar_dict)
        incar.check_params()
```

### Relevant files to reproduce this bug

Can be run using the minimal example

**Comment by wladerer** [comment]
Ah ok I see why this was not caught. In tests.io.vasp.inputs, the only test for Incar.check_params() is

```python 
    def test_check_params(self):
        # Triggers warnings when running into invalid parameters
        with pytest.warns(BadIncarWarning) as record:
            incar = Incar(
                {
                    "ADDGRID": True,
                    "ALGO": "Normal",
                    "AMIN": 0.01,
                    "ICHARG": 1,
                    "MAGMOM": [1, 2, 4, 5],
                    "NBAND": 250,  # typo in tag
                    "METAGGA": "SCAM",  # typo in value
                    "EDIFF": 5 + 1j,  # value should be a float
                    "ISIF": 9,  # value out of range
                    "LASPH": 5,  # value should be bool
                    "PHON_TLIST": "is_a_str",  # value should be a list
                }
            )
            incar.check_params()
```

Which means that no matter the incar input, it will catch a warning and pass.

**Comment by DanielYang59** [comment]
Hi @wladerer thanks for reporting. Can you please double check your `pymatgen` version? As the following code has been updated four months ago from:
```python 
if param_type is not None and not isinstance(val, eval(param_type)):
``` 

To: https://github.com/materialsproject/pymatgen/blob/720fdf86539236c737f12250ae71d68e3872c616/src/pymatgen/io/vasp/inputs.py#L1032

With the intention to avoid using of `eval` (perhaps I overreacted there as `param_type` would not come from untrusted source, and therefore should not introduce safety concerns).

Even with the latest version, it looks like there is indeed an issue with handling of type of `LREAL` here: https://github.com/materialsproject/pymatgen/blob/720fdf86539236c737f12250ae71d68e3872c616/src/pymatgen/io/vasp/incar_parameters.json#L652-L661

I just had another complete check of `incar_parameters.json` and it seems `LREAL` is the only tag that requires a Union type here, and I must have missed it, apologies.

I would fix this ASAP.

> Which means that no matter the incar input, it will catch a warning and pass.

The checking mechanism are those following: https://github.com/materialsproject/pymatgen/blob/720fdf86539236c737f12250ae71d68e3872c616/tests/io/vasp/test_inputs.py#L786-L791

**Comment by wladerer** [comment]
Hi I very unfortunately had a python environment disaster when trying to install a different package. I pulled that version I reported using the command line command recommended in the environment where I was developing in. I'm sorry that I am unable to double check :(

I will try and see if I have any way of tracing it back and recreating the env and will report back. 

And I see now that this test is sufficient since it only catches warnings rather than errors. Sorry for the confusion. 

Thanks!
